### ISSUE TITLE
Consider Publishing External requestIdleCallback Polyfill

### ISSUE THREAD

**Issue by sebmarkbage** [facebook/react#11330]
We require a high fidelity requestIdleCallback polyfill and pretty much all of them out there are not. We can always fallback to our internal fallback but that still means that others have to ship their own, which isn't cooperating with ours. (Even at FB we have two.)

Our built-in one is not yet implementing the `timeout` option but since we want to start using it, we probably should.

This can become even more interesting if a [priority queue option](https://github.com/w3c/requestidlecallback/issues/68) lands (if it is sufficient for us), since we can start polyfilling that option even in native requestIdleCallback implementations.

**Comment by Tom910** [comment]
I wrote the implementation of the launch of the async actions in the browser, which is based on the requestAnimationFrame. Has support for priorities and frames break for 16ms. May be useful =) https://github.com/Tom910/frame-scheduling

**Comment by acdlite** [comment]
I think @clemmy will work on this. Will need to update polyfill to support multiple callbacks.

**Comment by inf3rno** [comment]
I think this is not possible with the timeout option, but I'd need the same. I'll let you know if I find something promising.

**Comment by inf3rno** [comment]
Interactive readyState for very old browsers:
http://dean.edwards.name/weblog/2006/06/again/

For IE it adds a deferred script with document.write. In theory our script and readystatechange approach uses the same, but it does not work. I ready in a comment, that document.write must be used. I tried innerHTML and appendChild, but none of them worked. The document.write is not good, because moha writes that the page was reloaded and stops testing.

**Comment by inf3rno** [comment]
We could try behaviors in IE too, but it is not supported from IE 10: https://stackoverflow.com/questions/26339276/what-is-behavior-url-property-in-css

**Comment by inf3rno** [comment]
Appending an inline script works in IE:

```js
function fn (cb, cnt) {
	if (!cnt)
		cnt = 0;
	var body = document.body;
	var script = document.createElement("script");
	script.text = 'window.mycb1()';
	window.mycb1 = function () {
		cb();
		delete window.mycb1;
		body.removeChild(script);
		if (cnt<50)
			fn(cb, cnt + 1);
	};
	body.appendChild(script);
}
```

```
beforeunload, 114ms, 1, blank
pagehide, 121ms, 1, blank
unload, 127ms, 1, blank
unload + append inline script, 134ms, 1, blank
unload + append inline script, 141ms, 1, blank
unload + append inline script, 148ms, 1, blank
unload + append inline script, 155ms, 1, blank
unload + append inline script, 160ms, 1, blank
unload + append inline script, 167ms, 1, blank
unload + append inline script, 175ms, 1, blank
unload + append inline script, 182ms, 1, blank
unload + append inline script, 188ms, 1, blank
unload + append inline script, 195ms, 1, blank
unload + append inline script, 202ms, 1, blank
unload + append inline script, 209ms, 1, blank
unload + append inline script, 216ms, 1, blank
unload + append inline script, 223ms, 1, blank
unload + append inline script, 230ms, 1, blank
unload + append inline script, 237ms, 1, blank
unload + append inline script, 244ms, 1, blank
unload + append inline script, 250ms, 1, blank
unload + append inline script, 257ms, 1, blank
unload + append inline script, 263ms, 1, blank
unload + append inline script, 271ms, 1, blank
unload + append inline script, 277ms, 1, blank
unload + append inline script, 284ms, 1, blank
unload + append inline script, 290ms, 1, blank
x child script, 306ms, <--, 1, x.html
x child load, 318ms, 1, x.html
unload + MessageChannel, 324ms, 1, x.html
unload + requestAnimationFrame, 330ms, 1, x.html
unload + postMessage, 337ms, 1, x.html
unload + setTimeout 0, 343ms, 1, x.html
y child script, 3132ms, <--, 2, y.html
y child load, 3144ms, 2, y.html
```

It is called only 24 times instead of 50 and it is never called with a new document having a loading readyState, so it is not good for injection or adding event listeners.

**Comment by inf3rno** [comment]
Using a deferred or async script does not make a difference from this perspective. The defer means that it is loaded only after the parent page is loaded, which is. The async means that the external script is loaded asynchronously. The async is turned on by default except very old Firefox. If you turn it off, then it will load the script in a non-blocking but sync way. By inline scripts it does not make sense.

https://www.html5rocks.com/en/tutorials/speed/script-loading/

**Comment by inf3rno** [comment]
I tried out script with ready state change:

```js
var body = document.body;
var script = document.createElement("script");
script.type = "text/javascript";
script.src = "javascript:";
script.onreadystatechange = function () {
	delete script.onreadystatechange;
	body.removeChild(script);
	cb();
};
body.appendChild(script);
```

It fires after the child script in IE10, so it is not good, but it has the smallest delay. 

```
beforeunload, 118ms, 1, blank
unload, 126ms, L, 1, blank
unload + append inline script, 133ms, L, 1, blank
x child script, 144ms, <-[DL]-, 1, x.html
x child load, 155ms, D, 1, x.html
unload + script readystatechange, 888ms, D, 1, x.html
unload + MessageChannel, 894ms, D, 1, x.html
unload + requestAnimationFrame, 900ms, D, 1, x.html
unload + setTimeout 0, 906ms, D, 1, x.html
unload + setImmediate, 911ms, D, 1, x.html
unload + postMessage, 917ms, D, 1, x.html
y child script, 1646ms, <-[DL]-, 2, y.html
y child load, 1659ms, D, 2, y.html
```

Still the delay is huge and a few dozen msecs does not really matter.

**Comment by inf3rno** [comment]
Sorry, I messaged the previous ones to the wrong issue. Removed them. I checked all possible deferring solutions and for IE10, IE11 and PhantomJS there aren't really good solutions. My test was the following: I opened a child window with `window.open`. I loaded a page to the child window and deferred my callbacks from the `unload` of the child window.

```js
var win = window.open("about:blank", uuid, "menubar=yes,location=yes,resizable=yes,scrollbars=yes,status=yes");
win.addEventListener("unload", function () {
	deferAll(function (name) {
		console.log(name);
	});
});
win.location.href = "http://localhost:9876/base/test/client/x.html";
```

I used Karma to test this in Firefox, Opera, Chrome, PhantomJS, IE11 and emulated IE10. Some of the methods call the callback after unload but before the new document is created in the child window. I call these forced methods. The others are called after the new document was created. Some of them when the new document is in a loading readyState while others after the new document in the child window was loaded. I call the first group lightly forced, the second group delayed, because somehow it has a few hundred msec delay compared to the load event of the new document. For me the lightly forced would be very useful because I could inject script into the child document before the scripts in its head run, but sadly PhantomJS, IE11 and IE10 don't have lightly forced solutions. For you every forced or lightly forced solutions might be good to call the callback after the timeout:

The summary of the test results:
```
Chrome 68 + Karma
----------------- forced -----------------------
append inline script, 77ms
Promise.resolve, 77ms
-------------- lightly forced ------------------
postMessage, 92ms
MessageChannel, 93ms
setTimeout 0, 93ms
requestIdleCallback 1, 93ms
---------------- delayed -----------------------
requestIdleCallback, 98ms
requestIdleCallback 0.1, 99ms
requestIdleCallback 0, 99ms
requestAnimationFrame did not fire


Opera 54
----------------- forced -----------------------
append inline script, 187ms
Promise.resolve, 187ms
-------------- lightly forced ------------------
postMessage, 191ms
MessageChannel, 191ms
requestIdleCallback 1, 191ms
---------------- delayed -----------------------
requestIdleCallback, 206ms
requestIdleCallback 0.1, 206ms
requestIdleCallback 0, 206ms
setTimeout 0, 322ms
requestAnimationFrame did not fire


PhantomJS 2.1
----------------- forced -----------------------
append inline script, 357ms
-------------- lightly forced ------------------
---------------- delayed -----------------------
postMessage, 359ms
setTimeout 0, 359ms
requestAnimationFrame, 360ms
MessageChannel, 360ms


IE 10
----------------- forced -----------------------
append inline script, 505ms
-------------- lightly forced ------------------
---------------- delayed -----------------------
script readystatechange, 586ms
MessageChannel, 595ms
requestAnimationFrame, 602ms
setTimeout 0, 614ms
setImmediate, 624ms
postMessage, 646ms


IE 11
----------------- forced -----------------------
append inline script, 606ms
-------------- lightly forced ------------------
---------------- delayed -----------------------
MessageChannel, 672ms
postMessage, 679ms
requestAnimationFrame, 686ms
setTimeout 0, 693ms
setImmediate, 699ms


Firefox 61
----------------- forced -----------------------
append inline script, 148ms
Promise.resolve, 148ms
-------------- lightly forced ------------------
setTimeout 0, 157ms
requestIdleCallback 0.1, 157ms
requestIdleCallback 0, 157ms
requestIdleCallback 1, 158ms
postMessage, 158ms
MessageChannel, 159ms
requestAnimationFrame, 159ms
---------------- delayed -----------------------
requestIdleCallback, 280ms
```

Here are the defers:

```js
"setTimeout 0": {
	fn: function (cb) {
		setTimeout(cb, 0);
	}
},
"requestIdleCallback": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb);
	}
},
"requestIdleCallback 1": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb, {timeout: 1});
	}
},
"requestIdleCallback 0.1": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb, {timeout: 0.1});
	}
},
"requestIdleCallback 0": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb, {timeout: 0});
	}
},
"setImmediate": {
	deps: function () {
		return !!window.setImmediate;
	},
	fn: function (cb) {
		window.setImmediate(cb);
	}
},
"requestAnimationFrame": {
	deps: function () {
		return !!window.requestAnimationFrame;
	},
	fn: function (cb) {
		window.requestAnimationFrame(cb);
	}
},
"postMessage": {
	deps: function () {
		return !!window.postMessage;
	},
	fn: function (cb) {
		var mid = "pm." + Math.random();
		var origin = window.location.origin;
		if (!origin)
			origin = window.location.protocol + '//' + window.location.hostname + (window.location.port ? (':' + window.location.port) : '');
		// IE 10 does not support location.origin
		window.addEventListener("message", function handle(evt) {
			if (evt.source !== window || evt.origin !== origin || evt.data !== mid)
				return;
			window.removeEventListener("message", handle);
			evt.stopPropagation();
			cb();
		});
		window.postMessage(mid, origin);
	}
},
"MessageChannel": {
	deps: function () {
		return !!window.MessageChannel;
	},
	fn: function (cb) {
		var channel = new window.MessageChannel();
		channel.port1.onmessage = function (e) {
			cb();
		};
		channel.port2.postMessage(1);
	}
},
"Promise.resolve": {
	deps: function () {
		return !!window.Promise;
	},
	fn: function (cb) {
		window.Promise.resolve().then(cb);
	}
},
"append inline script": {
	fn: function (cb) {
		var body = document.body;
		var script = document.createElement("script");
		script.text = 'window.mycb1();';
		window.mycb1 = function () {
			cb();
			delete window.mycb1;
			body.removeChild(script);
		};
		body.appendChild(script);
	}
},
"script readystatechange": {
	deps: function () {
		return "onreadystatechange" in document.createElement("script");
	},
	fn: function (cb) {
		var body = document.body;
		var script = document.createElement("script");
		script.type = "text/javascript";
		script.src = "javascript:";
		script.onreadystatechange = function () {
			delete script.onreadystatechange;
			body.removeChild(script);
			cb();
		};
		body.appendChild(script);
	}
}
```

Good luck!

**Comment by inf3rno** [comment]
I checked again whether the forced ones are sync, and it turned out that the "append inline script" is sync. So you cannot use that one for next tick.

**Comment by gaearon** [comment]
We're not using requestIdleCallback anymore.
